src/mptcpize.c

```diff
@@ -0,0 +1,226 @@
+		help();
+		return -1;
+	}
+
+	// build environment, copying the current one ...
+	while (environ[nr])
+		nr++;
+	envp = calloc(nr + 3, sizeof(char *));
+	if (!envp)
+		error(1, errno, "can't allocate env list");
+
+	// ... filtering out any 'LD_PRELOAD' ...
+	nr = 0;
+	i = 0;
+	while (environ[nr]) {
+		if (strncmp(environ[nr], "LD_PRELOAD=", 11)) {
+			envp[i] = environ[nr];
+			i++;
+		}
+		nr++;
+	}
+
+	// ... appending the mptcpwrap preload...
+	envp[i++] = MPTCPWRAP_ENV;
+
+	// ... and enable dbg if needed
+	if (debug)
+		envp[i++] = "MPTCPWRAP_DEBUG=1";
+
+	// build the NULL terminated arg list
+	argv = calloc(argc + 1, sizeof(char *));
+	if (!argv)
+		error(1, errno, "can't allocate argument list");
+
+	memcpy(argv, av, argc * sizeof(char*));
+	return execvpe(argv[0], argv, envp);
+}
+
+#define SYSTEMCTL_SHOW "systemctl show -p FragmentPath "
+
+static char *locate_unit(const char *name)
+{
+	char *cmd, *line = NULL;
+	FILE *systemctl;
+	size_t len = 0;
+	ssize_t read;
+
+	if (!access(name, R_OK))
+		return strdup(name);
+
+	len = strlen(name) + 1 + strlen(SYSTEMCTL_SHOW);
+	cmd = malloc(len);
+	if (!cmd)
+		error(1, 0, "can't allocate systemctl command string");
+	sprintf(cmd, SYSTEMCTL_SHOW"%s", name);
+
+	systemctl = popen(cmd, "r");
+	if (!systemctl)
+		error(1, errno, "can't execute %s", cmd);
+
+	while ((read = getline(&line, &len, systemctl)) != -1) {
+		if (!strncmp(line, "FragmentPath=", 13)) {
+			char *ret = strdup(&line[13]);
+
+			// trim trailing newline, if any
+			len = strlen(ret);
+			if (len > 0 && ret[len - 1] == '\n')
+				ret[len - 1] = 0;
+			free(line);
+			pclose(systemctl);
+			return ret;
+		}
+	}
+
+	error(1, 0, "can't find FragmentPath attribute for unit %s", name);
+
+	// never reached: just silence gcc
+	return NULL;
+}
+
+static int unit_update(int argc, char *argv[], int enable)
+{
+	char *unit, *line = NULL;
+	char dst_path[PATH_MAX];
+	int append_env = enable;
+	size_t len = 0;
+	ssize_t read;
+	FILE *src;
+	int dst;
+
+	if (argc < 1) {
+		fprintf(stderr, "missing unit argument");
+		help();
+		return -1;
+	}
+
+	unit = locate_unit(argv[0]);
+	src = fopen(unit, "r");
+	if (!src)
+		error(1, errno, "can't open file %s", unit);
+
+	strcpy(dst_path, "/tmp/unit_XXXXXX");
+	dst = mkstemp(dst_path);
+	if (dst < 0)
+		error(1, errno, "can't create tmp file");
+
+	// reset any prior error, to allow later check on errno
+	errno = 0;
+
+	/**
+	 * systemd does not allow Environment property update via a command,
+	 * we have to copy the existing unit file to a tmp one
+	 */
+	while ((read = getline(&line, &len, src)) != -1) {
+		int is_env = !strncmp(line, "Environment=", 12);
+
+		if (!is_env) {
+			if (write(dst, line, read) < 0)
+				error(1, errno, "can't write '%s' into %s", line, dst_path);
+		}
+
+		if (append_env && (is_env || !strncmp(line, "[Service]", 9))) {
+			if (write(dst, "Environment=", 12) < 0 ||
+			    write(dst, MPTCPWRAP_ENV, strlen(MPTCPWRAP_ENV)) < 0 ||
+			    write(dst, "\n", 1) < 0)
```

Comment:
(no need to change but maybe clearer with `dprintf()` :) )

Reply:
ok, I'll use `dprintf()` in the next iteration. I did not know about it (!!!)